File: cpp_100-999/210.cc
// C++
// leetcode 210
// https://leetcode-cn.com/problems/course-schedule-ii/
// 小改了一下刚刚写的207的答案，同样是时空间都很差的屑方法

class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
    	bool iRelayj[numCourses][numCourses];
    	memset(iRelayj, 0, sizeof(iRelayj));
    	for(auto & p : prerequisites) {
    		iRelayj[p[0]][p[1]] = true;
    	}
    	bool finished[numCourses];
    	memset(finished, 0, sizeof(finished));
    	vector<int> done;
    	bool lastTurnLearnedSomething = true;
    	
    	while(done.size() < numCourses && lastTurnLearnedSomething) {
    		lastTurnLearnedSomething = false;
	    	for(int i = 0; i < numCourses; i++) {
	    		// 看一下课i能否学
	    		if(finished[i]) continue;	// 之前学过了
	    		bool canLearn = true;
	    		for(int j = 0; j < numCourses; j++) {
	    			if(iRelayj[i][j] == true) {
	    				canLearn = false;
	    				break;
	    			}
	    		}
	    		if(canLearn) {
	    			// 学课i
	    			lastTurnLearnedSomething = true;
	    			done.push_back(i);
	    			finished[i] = true;
	    			for(int j = 0; j < numCourses; j++) {
	    				iRelayj[j][i] = false;
	    			}
	    		}
	    	}
	    }
	    if(done.size() < numCourses) done.clear();
	    return done;
    }
};

File: Go/Go_100-999/210.go
// Go
// leetcode 210
// https://leetcode.cn/problems/course-schedule-ii/

func findOrder(n int, prerequisites [][]int) []int {
	taken := make([]bool, n)
	res := make([]int, 0, n)
	next := make([][]int, n)
	inDeg := make([]int, n)
	for _, p := range prerequisites {
		a, b := p[1], p[0] // 上完 a 才能上 b
		next[a] = append(next[a], b)
		inDeg[b]++
	}

	for len(res) < n {
		cur := 0 // 本轮上的课的数量

		for i:=0; i<n; i++ {
			if taken[i] { // 上过
				continue
			}

			if inDeg[i] != 0 { // 需要前置课程
				continue
			}

			taken[i] = true
			cur++
			res = append(res, i)
			for _, x := range next[i] {
				inDeg[x]--
			}
		}
		if cur == 0 {
			break
		}
	}

	if len(res) != n {
		return []int{}
	}

	return res
}